ciri/node/network/router.c

```diff
@@ -0,0 +1,378 @@
+    return RC_FAILED_HANDSHAKE;
+  }
+
+  // Check whether the same coordinator address is used
+  if (memcmp(router->node->conf.coordinator_address, handshake->coordinator_address,
+             HANDSHAKE_COORDINATOR_ADDRESS_BYTES_LENGTH) != 0) {
+    log_warning(logger_id, "Failed handshake with tcp://%s:%d because different coordinator addresses are used\n", ip,
+                port);
+    return RC_FAILED_HANDSHAKE;
+  }
+
+  // Check whether the same MWM is used
+  if (router->node->conf.mwm != handshake->mwm) {
+    log_warning(logger_id, "Failed handshake with tcp://%s:%d because different MWM values are used: %d and %d\n", ip,
+                port, router->node->conf.mwm, handshake->mwm);
+    return RC_FAILED_HANDSHAKE;
+  }
+
+  // Check whether we support the supported protocol versions by the neighbor
+  if ((protocol_version = handshake_supported_version(
+           handshake_supported_protocol_versions, sizeof(handshake_supported_protocol_versions),
+           handshake->supported_versions, header_length - HANDSHAKE_MIN_BYTES_LENGTH + 1)) < 0) {
+    log_warning(logger_id, "Failed handshake with tcp://%s:%d because of protocol version %d mismatch\n", ip, port,
+                protocol_version);
+    return RC_FAILED_HANDSHAKE;
+  }
+
+  // Check whether the peer is an allowed neighbor
+  if ((*neighbor = router_neighbor_find_by_endpoint_values(router, ip, handshake_port)) == NULL) {
+    log_warning(logger_id, "Failed handshake with tcp://%s:%d because the peer is a non-tethered neighbor\n", ip, port);
+    return RC_FAILED_HANDSHAKE;
+  }
+  // TODO auto-peering
+
+  // Check if the neighbor is not already connected
+  if ((*neighbor)->endpoint.stream != NULL) {
+    log_warning(logger_id, "Failed handshake with tcp://%s:%d because it's an already connected neighbor\n", ip, port);
+    return RC_FAILED_HANDSHAKE;
+  }
+
+  (*neighbor)->protocol_version = protocol_version;
+
+  return RC_OK;
+}
+
+retcode_t router_neighbor_read(router_t *const router, neighbor_t *const neighbor, void const *const buffer,
+                               size_t const buffer_size) {
+  retcode_t ret = RC_OK;
+  protocol_header_t const *header = NULL;
+  uint16_t header_length = 0;
+  size_t chunk_size = 0;
+
+  if (router == NULL || neighbor == NULL || buffer == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  if (buffer_size == 0) {
+    return RC_OK;
+  }
+
+  // We concatenate read data to previously read data
+  chunk_size = MIN(PACKET_MAX_BYTES_LENGTH - neighbor->buffer_size, buffer_size);
+  memcpy(neighbor->buffer + neighbor->buffer_size, buffer, chunk_size);
+  neighbor->buffer_size += chunk_size;
+
+  // We haven't received a full header yet
+  if (neighbor->buffer_size <= HEADER_BYTES_LENGTH) {
+    return RC_OK;
+  }
+
+  header = (protocol_header_t const *)neighbor->buffer;
+  header_length = ntohs(header->length);
+
+  // We haven't received the full packet yet
+  if (neighbor->buffer_size - HEADER_BYTES_LENGTH < header_length) {
```

Comment:
should it not be `if (neighbor->buffer_size < HEADER_BYTES_LENGTH)`?
i don't understand why you compare the size of the part of buffer which is not header to header..

Reply:
`header_length` contains the advertised length in the header. It indicates how much data is coming after. So in the buffer we need to have at least this amount.